### PR TITLE
ARROW-10932: [C++] BinaryMemoTable::CopyOffsets access out-of-bound address when data is empty

### PR DESCRIPTION
**Pull request by Yuhta**

### CODE REVIEW

**Review by pitrou — APPROVED** [review]
+1, thank you @Yuhta !

**Comment by pitrou** [comment]
The CI failures are unrelated, will merge.

File: cpp/src/arrow/util/hashing.h [hunk 1]
```diff
@@ -690,7 +690,8 @@ class BinaryMemoTable : public MemoTable {
     DCHECK_LE(start, size());
 
     const builder_offset_type* offsets = binary_builder_.offsets_data();
-    const builder_offset_type delta = offsets[start];
+    const builder_offset_type delta =
+        start < binary_builder_.value_data_length() ? offsets[start] : 0;
```
Comment (Yuhta) [inline cpp/src/arrow/util/hashing.h:694]: I made a mistake here, `binary_builder_.value_data_length()` should be `binary_builder_.length()`

Comment (pitrou) [reply cpp/src/arrow/util/hashing.h:694]: Can you post a followup PR (just reuse the same JIRA number)?

**Pull request closed without merging**